### PR TITLE
Add Akima1DInterpolator to cupyx.scipy.interpolate

### PR DESCRIPTION
**Pull request by andfoy**
See https://github.com/cupy/cupy/issues/7186

### CODE REVIEW

**Comment by ev-br** [comment]
https://github.com/andfoy/cupy/pull/1 ports one more test for an edge case found in a recent scipy release cycle: in scipy <=1.9.x, e.g. Akima1DInterpolator was not very graceful in handling zero size `y` arrays. The fix should be a simple addition of  `initial=-np.inf` to a `max` call: https://github.com/scipy/scipy/pull/17305/files#diff-7aed71efd6974c5ff12f00cca2883cbf7f3a72bb5efc50e97145266b3b3eebd3R442

**Comment by andfoy** [comment]
@takagi @ev-br, I think we need first to implement the `initial` kwarg to `max/min`

**Comment by ev-br** [comment]
An alternative workaround could the be to and a 
explicit shotcut if `y.size==0`:

https://github.com/scipy/scipy/blob/main/scipy/interpolate/_cubic.py#L646

File: cupyx/scipy/interpolate/_bspline2.py [hunk 1]
```diff
@@ -554,7 +554,8 @@ def _make_periodic_spline(x, y, t, k, axis):
     extradim = prod(y.shape[1:])
     rhs = cupy.empty((n + k - 1, extradim), dtype=float)
     rhs[:(k - 1), :] = 0
-    rhs[(k - 1):, :] = y.reshape((-1, extradim))
+    rhs[(k - 1):, :] = (y.reshape(n, 0) if y.size == 0 else
+                        y.reshape((-1, extradim)))
```
Comment (andfoy) [inline cupyx/scipy/interpolate/_bspline2.py:558]: @ev-br, I had to do this modification in order to fix the `test_zero_size`, which otherwise would fail with the following stacktrace:

```python
Traceback (most recent call last):
  File "/home/andfoy/Documentos/Quansight/cupy/cupy/testing/_loops.py", line 45, in _call_func
    result = impl(*args, **kw)
  File "/home/andfoy/Documentos/Quansight/cupy/tests/cupyx_tests/scipy_tests/interpolate_tests/test_polyint.py", line 389, in test_zero_size
    obj = cls(x, y, bc_type=bc_type)
  File "/home/andfoy/Documentos/Quansight/cupy/cupyx/scipy/interpolate/_bspline2.py", line 251, in make_interp_spline
    return _make_periodic_spline(x, y, t, k, axis)
  File "/home/andfoy/Documentos/Quansight/cupy/cupyx/scipy/interpolate/_bspline2.py", line 557, in _make_periodic_spline
    rhs[(k - 1):, :] = y.reshape((-1, extradim))
  File "cupy/_core/core.pyx", line 765, in cupy._core.core._ndarray_base.reshape
    return _manipulation._ndarray_reshape(self, shape, order)
  File "cupy/_core/_routines_manipulation.pyx", line 83, in cupy._core._routines_manipulation._ndarray_reshape
    return _reshape(self, shape)
  File "cupy/_core/_routines_manipulation.pyx", line 352, in cupy._core._routines_manipulation._reshape
    shape = internal.infer_unknown_dimension(shape_spec, self.size)
  File "cupy/_core/internal.pyx", line 208, in cupy._core.internal.infer_unknown_dimension
    raise ValueError(
ValueError: cannot reshape array of size 0 into shape (-1, 0)
```

**Comment by takagi** [comment]
Record: Same as PchipInterpolator, Akima1DInterpolator inherits from CubicHermiteSpline and PPoly, we can refer its performance comparison to https://github.com/cupy/cupy/pull/7204#issuecomment-1352117803.

**Review by takagi — APPROVED** [review]
LGTM!

File: cupyx/scipy/interpolate/_cubic.py [hunk 2]
```diff
@@ -318,3 +320,95 @@ def pchip_interpolate(xi, yi, x, der=0, axis=0):
...
+        ind = cupy.nonzero(f12 > 1e-9 * max_value)
+        x_ind, y_ind = ind[0], ind[1:]
+        # Set the slope at breakpoint
+        t[ind] = (f1[ind] * m[(x_ind + 1,) + y_ind] +
+                  f2[ind] * m[(x_ind + 2,) + y_ind]) / f12[ind]
+
+        super().__init__(x, y, t, axis=0, extrapolate=False)
+        self.axis = axis
+
+    def extend(self, c, x, right=True):
```
[the remark this answers is not in the document]
Comment (ev-br) [reply cupyx/scipy/interpolate/_cubic.py:400]: This stub comes straight ftim scipy: the method is inherited from PPoly, and the default implementation makes no sense for Akima1D spline.

[the remark this answers is not in the document]
Comment (takagi) [reply cupyx/scipy/interpolate/_cubic.py:400]: Ah, okay. Thanks!

**Comment by takagi** [comment]
Thanks, @andfoy and @ev-br!
